### PR TITLE
Show path in recent

### PR DESCRIPTION
**Pull request by jeremypw**
Because items in recent:// 'folder' are allowed to have the same name, this PR disambiguates them by using the path (up to a maximum length) as the display name.

Also SearchResults is modified so that the original folder of files is displayed, to disambiguate them.  Clicking on a search result now navigates to the actual file, not the item in recent://

TO TEST:

* If you already have files in your Recent folder then you should see the effect straight away in ListView and SearchBar.  
* Create two files with the same (best to be unusual) name in different folders which also have the same name then search for that file name.  You will see that those search results are distinguishable now.
* If you have History 'off' in Switchboard Privacy plug, then instead of Recent view you will get a Welcome view with an appropriate message and an opportunity to open Switchboard Privacy plug.
* Create two files as above then edit them with an app that records history (note Code does not seem to do this for some reason). Or play appropriately named videos or music files (depending on your settings - these will also appear in Recent).

You can run the PR from the build directory without installing.  Run master Files from Plank for comparison (one at a time).

Needs UX team opinion on new appearance of Recent VIew.

### CODE REVIEW

**Comment by jeremypw** [comment]
Example of new Recent View:

screenshot from 2018-11-13 13-08-42

**Comment by jeremypw** [comment]
Example of Recent View with History turned off:

screenshot from 2018-11-13 13 15 54

**Comment by naaando** [comment]
The limited_length_path method is returning ".../baz/abc.txt" if the limit length is 7 on a path "/foo/bar/baz/abc.txt", is that the desired behavior? I made some tests (#811) that aren't passing due to this.

Everything else, code looks good.

File: libcore/FileUtils.vala [hunk 1]
```diff
@@ -78,6 +78,58 @@ namespace PF.FileUtils {
         return parent_path;
     }
 
+    public const string ELLIPSIS = "…";
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply libcore/FileUtils.vala]: Fixed

File: libcore/FileUtils.vala [hunk 2]
```diff
@@ -78,6 +78,58 @@ namespace PF.FileUtils {
...
+
+        string path, protocol;
+        split_protocol_from_path (full_path, out protocol, out path);
+
+        string[] tokens = path.strip ().split (Path.DIR_SEPARATOR_S, 10);
+        int n_tokens = tokens.length;
+        bool has_protocol = protocol.length > 0;
+
+        if (has_protocol) {
+            if (max_length - protocol.length < 12) {
```
[the remark this answers is not in the document]
Comment (jeremypw) [reply libcore/FileUtils.vala]: Fixed

**Comment by jeremypw** [comment]
An unresolved issue is how to handle RTL locales.

**Comment by jeremypw** [comment]
Another question: After returning to this PR after some time, I am not sure whether the limit_length_path function is essential.  I think it was put in mainly so icon view did not have unwieldy names but an alternative solution would be for recent:// to be forced to show in ListView, where very long names are not a problem.

**Comment by jeremypw** [comment]
There are two more 'arbitrary' constants to be considered:
In GOF.File, the maximum length of the customary name is limited to '50' and in SearchView it is '35'.  I think the '50' was to keep names in IconView to 5 lines or less.

**Comment by jeremypw** [comment]
Closing as no approval after 11 months

**Comment by cassidyjames** [comment]
@jeremypw I would be in favor of showing the whole path in the overlay on hover rather than in the name itself. If you're interested in picking this back up. :grimacing:

**Comment by jeremypw** [comment]
I'll open an issue about it and maybe produce a new PR later.

**Pull request closed without merging**